bee-crypto/src/ternary/sponge/curlp.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::ternary::{sponge::Sponge, HASH_LENGTH};
+
+use bee_ternary::{Btrit, TritBuf, Trits};
+
+use std::{
+    convert::Infallible,
+    ops::{Deref, DerefMut},
+};
+
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+const HALF_STATE_LENGTH: usize = STATE_LENGTH / 2;
+
+const TRUTH_TABLE: [[Btrit; 3]; 3] = [
+    [Btrit::PlusOne, Btrit::Zero, Btrit::NegOne],
+    [Btrit::PlusOne, Btrit::NegOne, Btrit::Zero],
+    [Btrit::NegOne, Btrit::PlusOne, Btrit::Zero],
+];
+
+/// Available round numbers for CurlP.
+#[derive(Copy, Clone)]
+pub enum CurlPRounds {
+    /// 27 rounds.
+    Rounds27 = 27,
+    /// 81 rounds.
+    Rounds81 = 81,
+}
+
+/// State of the ternary cryptographic function `CurlP`.
+pub struct CurlP {
+    /// The number of rounds of hashing to apply to the state.
+    rounds: CurlPRounds,
+    /// The internal state.
+    state: TritBuf,
+    /// Workspace for performing transformations.
+    work_state: TritBuf,
+}
+
+impl CurlP {
+    /// Create a new `CurlP` sponge with `rounds` of iterations.
+    pub fn new(rounds: CurlPRounds) -> Self {
+        Self {
+            rounds,
+            state: TritBuf::zeros(STATE_LENGTH),
+            work_state: TritBuf::zeros(STATE_LENGTH),
+        }
+    }
+
+    /// Transforms the internal state of the `CurlP` sponge after the input was copied into the internal state.
+    ///
+    /// The essence of this transformation is the application of a substitution box to the internal state, which happens
+    /// `rounds` number of times.
+    fn transform(&mut self) {
+        #[inline]
+        fn truth_table_get(state: &Trits, p: usize, q: usize) -> Btrit {
+            // Safe to unwrap since indexes come from iteration on the state.
+            TRUTH_TABLE[state.get(q).unwrap() as usize + 1][state.get(p).unwrap() as usize + 1]
```

Comment:
in this line is an error that leads to `thread 'main' panicked at 'attempt to add with overflow'`

Reply:
Never saw that. How can I reproduce ?

Reply:
it can be reproduced by:
```
let tx = BundledTransaction::from_trits(&TritBuf::<T1B1Buf>::zeros(BundledTransaction::trit_len())).unwrap();
let mut trit_buf = TritBuf::<T1B1Buf>::zeros(BundledTransaction::trit_len());
tx.into_trits_allocated(&mut trit_buf);
let _ = CurlP81::new().digest(&trit_buf).unwrap();
````

Reply:
Looking at the code, it seems like this is a bit of a problem. `&Trits` contains balanced trits. The cast from `Btrit::NegOne` to `usize` will probably underflow to the maximum value of a `usize`: adding `1` to this value then causes underflow. Why the error occurred there and not in the original `usize` cast I don't know, but regardless: this code smells fishy. Did you mean to use unbalanced trits?

Reply:
Fixed